toolchain/check/import_cpp.cpp

```diff
@@ -1278,42 +1279,12 @@ static auto CreateFunctionParamsInsts(Context& context, SemIR::LocId loc_id,
            .call_params_id = call_params_id}};
 }
 
-// Imports a function declaration from Clang to Carbon. If successful, returns
-// the new Carbon function declaration `InstId`. If the declaration was already
-// imported, returns the mapped instruction.
-static auto ImportFunctionDecl(Context& context, SemIR::LocId loc_id,
-                               clang::FunctionDecl* clang_decl)
-    -> SemIR::InstId {
-  // Check if the declaration is already mapped.
-  if (SemIR::InstId existing_inst_id =
-          LookupClangDeclInstId(context, clang_decl);
-      existing_inst_id.has_value()) {
-    return existing_inst_id;
-  }
-
-  if (clang_decl->isVariadic()) {
-    context.TODO(loc_id, "Unsupported: Variadic function");
-    MarkFailedDecl(context, clang_decl);
-    return SemIR::ErrorInst::InstId;
-  }
-
-  if (clang_decl->getTemplatedKind() ==
-      clang::FunctionDecl::TK_FunctionTemplate) {
-    context.TODO(loc_id, "Unsupported: Template function");
-    MarkFailedDecl(context, clang_decl);
-    return SemIR::ErrorInst::InstId;
-  }
-  CARBON_CHECK(clang_decl->getFunctionType()->isFunctionProtoType(),
-               "Not Prototype function (non-C++ code)");
-
-  if (auto* method_decl = dyn_cast<clang::CXXMethodDecl>(clang_decl)) {
-    if (method_decl->isVirtual()) {
-      context.TODO(loc_id, "Unsupported: Virtual function");
-      MarkFailedDecl(context, clang_decl);
-      return SemIR::ErrorInst::InstId;
-    }
-  }
-
+// Creates a partial `FunctionDecl` and a `Function` without C++ thunk
+// information. Returns std::nullopt on failure. On success,
+// `FinishImportFunctionDecl()` must be later called.
+static auto StartImportFunctionDecl(Context& context, SemIR::LocId loc_id,
+                                    clang::FunctionDecl* clang_decl)
+    -> std::optional<std::tuple<SemIR::FunctionDecl, SemIR::Function>> {
```

Comment:
It's a bit unusual to pass around a SemIR entity type like `SemIR::Function` by value. Instead, it'd be more conventional to add the function to the `context.functions()` list in this function and return an ID to it. (That's sort of analogous to allocating it with `new` here and  returning a pointer to it.)

I think if you did that, you could also finish building the `FunctionDecl` entirely as part of this function, removing the need for a separate `FinishImportFunctionDecl` function, and for returning the `FunctionDecl` -- so this function can return just a `FunctionId` (or `None` on failure).

Reply:
Done.
I was surprised to see that we could mutate the `SemIR::Function` object after it was inserted and after `ReplaceInstBeforeConstantUse()` was called.